File: src/env/worker.py
import gym
import numpy as np
from copy import deepcopy
from torch.multiprocessing import Process


class Worker(Process):
    """Environment worker."""

    def __init__(self, idx, num_envs, channel, config):
        super(Worker, self).__init__()
        self.idx = idx
        self.num_envs = num_envs
        self.channel = channel
        self.is_multiplayer = config["num_players"] > 1

        # Create environments
        self.envs = []
        for _ in range(num_envs):
            if isinstance(config["env"], str):
                self.envs.append(gym.make(config["env"]))
            else:
                self.envs.append(deepcopy(config["env"]))
        for i in range(num_envs):
            self.envs[i].seed(i+num_envs*idx + config["seed"])

    def run(self):
        while True:
            msg = self.channel.recv()
            if msg["command"] == "step":
                self.channel.send(self.step(msg["act"]))
            elif msg["command"] == "reset":
                self.channel.send(self.reset())
            elif msg["command"] == "copy envs":
                self.channel.send(deepcopy(self.envs))
            elif msg["command"] == "close":
                break

        self.close()

    def reset(self):
        obs = []
        pid = []
        legal_act = []
        for env in self.envs:
            o, i = env.reset()
            obs.append(o)
            pid.append(i["pid"])
            legal_act.append(i["legal_act"])
        return np.array(obs), {"pid": pid, "legal_act": legal_act}

    def step(self, acts):
        obs_next_list, rew_list, done_list, pid, legal_act_list = [], [], [], [], []

        for i in range(self.num_envs):
            # Env step
            obs_next, rew, done, info = self.envs[i].step(acts[i])

            if done:
                obs_next, info = self.envs[i].reset()

            # Append
            obs_next_list.append(obs_next)
            rew_list.append(rew)
            done_list.append(done)
            if self.is_multiplayer:
                pid.append(info["pid"])
            else:
                pid.append(0)
            legal_act_list.append(info["legal_act"])
        obs_next = np.stack(obs_next_list, axis=0)
        rew = np.stack(rew_list, axis=0)
        done = np.stack(done_list, axis=0)

        legal_act = legal_act_list
        info = {"pid": pid, "legal_act": legal_act}

        return obs_next, rew, done, info

    def close(self):
        for env in self.envs:
            env.close()

File: gym/hex/src/gym/worker.rs

use std::thread;
use core_affinity::CoreId;
use crossbeam::channel::{Sender, Receiver};
use numpy::ndarray::{Array, Ix1};
use crate::gym::{Action, Obs, Info, Env, Episode};


// Worker Messages
#[derive(Debug)]
pub enum WorkerMessageIn {
    Reset{eid: usize},
    Step{act: Action, eid: usize, dist: Array<f32, Ix1>, pol_id: usize},
    Render{eid: usize},
    Close{eid: usize},
    Shutdown,
}
#[derive(Clone, Debug)]
pub struct WorkerMessageOut {
    pub obs: Obs,
    pub rew: Option<f32>,
    pub done: Option<bool>,
    pub info: Info,
    pub eid: usize,
}

/// Worker thread for Envs
pub struct Worker {
    thread: Option<thread::JoinHandle<()>>,
}

impl Worker {
    pub fn new(id: usize, num_envs_per_worker: usize, in_channel: Receiver<WorkerMessageIn>, out_channel: Sender<WorkerMessageOut>, eps_in: Sender<Episode>, gamma: f32, max_len: usize, core_id: Option<CoreId>, size: u8) -> Worker {
        let eid_start = id * num_envs_per_worker;
        let mut episodes = vec![Some(Episode::new(max_len)); num_envs_per_worker];

        let thread = thread::spawn(move || {
            // Set core affinity
            if let Some(core_id) = core_id {
                core_affinity::set_for_current(core_id);
            }

            let mut envs = vec![Env::new(size); num_envs_per_worker];

            loop {
                if let Ok(message) = in_channel.recv() {
                    match message {
                        WorkerMessageIn::Reset{eid} => {
                            // Reset
                            let local_eid = eid - eid_start;
                            let (obs, info) = envs[local_eid].reset();
                            let msg = WorkerMessageOut{
                                obs: obs.clone(),
                                rew: None,
                                done: None,
                                info: info.clone(),
                                eid,
                            };

                            // Send
                            if out_channel.try_send(msg).is_err() {
                                panic!("Error sending message to master");
                            }

                            // Collect
                            let episode = episodes[local_eid].as_mut().expect("Reset episode is None, but should always be some");
                            episode.clear();
                            episode.obs.push(obs);
                            episode.pid.push(info.pid);
                            episode.legal_act.push(info.legal_act);
                        },
                        WorkerMessageIn::Step{act, eid, dist, pol_id} => {
                            // Step
                            let local_eid = eid - eid_start;
                            let (obs, rew, done, info) = envs[local_eid].step(act);

                            if done {
                                let (next_obs, next_info) = envs[local_eid].reset();

                                // Send
                                if out_channel.try_send(WorkerMessageOut{
                                    obs: next_obs.clone(),
                                    rew: Some(rew.clone()),
                                    done: Some(done.clone()),
                                    info: next_info.clone(),
                                    eid,
                                }).is_err() {
                                    panic!("Error sending message to master");
                                }

                                // Take episode, send to process and create new one
                                let mut episode = episodes[local_eid].take().expect("Step episode is None, but should always be some");
                                episode.act.push(act);
                                episode.dist.push(dist);
                                episode.rew.push(rew);
                                episode.done.push(done);
                                episode.pol_id.push(pol_id);

                                episode.process(gamma);
                                if eps_in.try_send(episode).is_err() {
                                    panic!("Error sending episode to master");
                                }

                                let mut new_episode = Episode::new(max_len);
                                new_episode.obs.push(next_obs);
                                new_episode.pid.push(next_info.pid);
                                new_episode.legal_act.push(next_info.legal_act);
                                episodes[local_eid] = Some(new_episode);
                            } else {
                                // Send
                                if out_channel.try_send(WorkerMessageOut{
                                    obs: obs.clone(),
                                    rew: Some(rew.clone()),
                                    done: Some(done.clone()),
                                    info: info.clone(),
                                    eid,
                                }).is_err() {
                                    panic!("Error sending message to master");
                                }

                                let episode = episodes[local_eid].as_mut().expect("Step episode is None, but should always be some");
                                episode.obs.push(obs);
                                episode.act.push(act);
                                episode.dist.push(dist);
                                episode.rew.push(rew);
                                episode.done.push(done);
                                episode.pol_id.push(pol_id);
                                episode.pid.push(info.pid);
                                episode.legal_act.push(info.legal_act);
                            }
                        },
                        WorkerMessageIn::Render{eid} => {
                            // Render
                            let local_eid = eid - eid_start;
                            println!("Render Env ID: {}\n{}", eid, envs[local_eid].to_string());
                        },
                        WorkerMessageIn::Close{eid} => {
                            // Close
                            let local_eid = eid - eid_start;
                            envs[local_eid].close();
                        },
                        WorkerMessageIn::Shutdown => {
                            for env in envs.iter_mut() {
                                env.close();
                            }
                            break;
                        },
                    }
                }
            }
        });

        Worker {
            thread: Some(thread),
        }
    }

    pub fn close(&mut self) {
        if let Some(handle) = self.thread.take() {
            handle.join().expect("Could not join env worker thread!");
        }
    }
}